src/darwin/Framework/CHIPTests/MTROTAProviderTests.m

```diff
@@ -864,7 +898,437 @@ - (void)test005_DoBDXTransferAllowUpdateRequest
+            MTROTASoftwareUpdateProviderClusterApplyUpdateRequestParams * params, ApplyUpdateRequestCompletion completion) {
+            XCTAssertEqualObjects(nodeID, @(kDeviceId2));
+            XCTAssertEqual(controller, sController);
+            XCTAssertEqualObjects(params.updateToken, updateToken1);
+            XCTAssertEqualObjects(params.newVersion, kUpdatedSoftwareVersion1); // TODO: Factor this out better!
+
+            XCTAssertTrue([[NSFileManager defaultManager] contentsEqualAtPath:otaRawImagePath2 andPath:kOtaDownloadedFilePath2]);
+
+            sOTAProviderDelegate.applyUpdateRequestHandler = nil;
+            [sOTAProviderDelegate respondToApplyUpdateRequestWithAction:MTROTASoftwareUpdateProviderOTAApplyUpdateActionProceed
+                                                             completion:completion];
+            [applyUpdateRequestExpectation2 fulfill];
+        };
+
+        sOTAProviderDelegate.notifyUpdateAppliedHandler = ^(NSNumber * nodeID, MTRDeviceController * controller,
+            MTROTASoftwareUpdateProviderClusterNotifyUpdateAppliedParams * params, MTRStatusCompletion completion) {
+            XCTAssertEqualObjects(nodeID, @(kDeviceId2));
+            XCTAssertEqual(controller, sController);
+            XCTAssertEqualObjects(params.updateToken, updateToken1);
+            XCTAssertEqualObjects(params.softwareVersion, kUpdatedSoftwareVersion1);
+
+            sOTAProviderDelegate.notifyUpdateAppliedHandler = nil;
+            [sOTAProviderDelegate respondSuccess:completion];
+            [notifyUpdateAppliedExpectation2 fulfill];
+        };
+    };
+
+    // Advertise ourselves as an OTA provider.
+    XCTestExpectation * announceResponseExpectation1 = [self announceProviderToDevice:device1];
+
+    // Make sure we get our callbacks in order.  Give it a bit more time, because
+    // we want to allow time for the BDX download.
+    [self waitForExpectations:@[ queryExpectation1, bdxBeginExpectation1, bdxQueryExpectation1, bdxEndExpectation1 ]
+                      timeout:(kTimeoutWithUpdateInSeconds) enforceOrder:YES];
+
+    // Nothing really defines the ordering of bdxEndExpectation and
+    // applyUpdateRequestExpectation with respect to each other.
+    [self waitForExpectations:@[ applyUpdateRequestExpectation1, notifyUpdateAppliedExpectation1 ]
+                      timeout:kTimeoutInSeconds
+                 enforceOrder:YES];
+
+    // Nothing defines the ordering of announceResponseExpectation with respect
+    // to _any_ of the above expectations.
+    [self waitForExpectations:@[ announceResponseExpectation1, announceResponseExpectation2 ] timeout:kTimeoutInSeconds];
+
+    // Make sure we get our callbacks in order.  Give it a bit more time, because
+    // we want to allow time for the BDX download.
+    [self waitForExpectations:@[
+        queryExpectation2, queryExpectation3, bdxBeginExpectation2, bdxQueryExpectation2, bdxEndExpectation2
```

Comment:
So... some of these are supposed to be ordered in some way with bdxBeginExpectation1 and whatnot, no?  I guess there is no good way to enforce that here for queryExpectation2, but maybe worth a comment about why we are not trying to enforce it.

And we can _absolutely_ check that queryExpectation3 comes after bdxEndExpectation1, since we don't set that handler up until we fulfill bdxEndExpectation1.  So everything here except queryExpectation2 should move into that first waitForExpectations that has queryExpectation1 and whatnot.

Reply:
done.